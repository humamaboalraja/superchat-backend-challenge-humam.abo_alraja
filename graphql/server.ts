import { ApolloServer } from 'apollo-server-express';
import express from 'express';
import { createServer } from 'http';
import { createContext } from './context';
import { schema } from './schema';

const { PORT = 5000 } = process.env;

const app = express();

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: process.env.STAGE !== 'production',
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
