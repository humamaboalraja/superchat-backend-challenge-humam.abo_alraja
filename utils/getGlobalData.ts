import { prisma } from '../singletons/prisma';

export default async function getGlobalData() {
  const messages = await prisma.message.findMany({});
  const contacts = await prisma.contact.findMany({});
  if (!messages || !contacts) throw new Error('Unable to set up global data');
  return { messages, contacts };
}
