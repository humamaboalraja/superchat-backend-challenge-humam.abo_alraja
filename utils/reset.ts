import { prisma } from '../singletons/prisma';

export default async function reset() {
  try {
    const models: string[] = ['Message', 'Contact'];
    for (const model in models) {
      await prisma.$executeRaw(`DELETE FROM "${models[model]}";`);
    }
  } catch (e) {
    throw new Error(`${e}`);
  }
}
