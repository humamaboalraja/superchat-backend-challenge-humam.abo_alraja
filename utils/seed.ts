import cryptoRandomString from 'crypto-random-string';
import { prisma } from '../singletons/prisma';

export default async function seed() {
  try {
    const contact1 = await prisma.contact.create({
      data: {
        name: 'Humam Abo Alraja',
        email: 'humam@company.com',
      },
    });
    const contact2 = await prisma.contact.create({
      data: {
        name: 'John',
        email: 'john@company.com',
      },
    });

    const contact3 = await prisma.contact.create({
      data: {
        name: 'Emma Mockassin',
        email: 'emma@company.com',
      },
    });

    const contact4 = await prisma.contact.create({
      data: {
        name: 'Alex Smith',
        email: 'alex@company.com',
      },
    });

    if (!contact1 || !contact2 || !contact3 || !contact4)
      throw new Error('no contacts');

    const one = await prisma.message.create({
      data: {
        messageContent: 'Hey Emma, How is it going?!',
        sender: {
          connect: {
            id: contact1.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });

    const two = await prisma.message.create({
      data: {
        messageContent: "Hi Humam, it's going great! How about you?",
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact1.id,
          },
        },
      },
    });

    const three = await prisma.message.create({
      data: {
        messageContent:
          'Hey Alex, do you mind sending me the pitch ASAP? Thanks in advance!',
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact4.id,
          },
        },
      },
    });

    const four = await prisma.message.create({
      data: {
        messageContent: 'Hi Emma, Definitely!',
        sender: {
          connect: {
            id: contact4.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });
  } catch (e) {
    throw new Error(`${e}`);
  }
}
