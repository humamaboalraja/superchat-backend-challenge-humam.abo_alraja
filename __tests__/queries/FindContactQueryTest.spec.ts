import getGlobalData from '../../utils/getGlobalData';
import reset from '../../utils/reset';
import seed from '../../utils/seed';
import server from '../server';

beforeAll(async () => {
  await reset();
  await seed();
  global.testData = await getGlobalData();
});

const FIND_CONTACT_QUERY = `
  query FindContact($email: String) {
   findContact(email: $email) {
     name
     email
   }
 }
  `;

describe('Test find contact query', () => {
  it('should find a contact given their email', async () => {
    const res = await server.executeOperation({
      query: FIND_CONTACT_QUERY,
      variables: { email: global.testData.contacts[0].email },
    });
    console.log(global.testData.contacts[0].email);
    expect(res).toMatchSnapshot();
  });

  it('should not find a contact if no arguments are passed in', async () => {
    const res = await server.executeOperation({
      query: FIND_CONTACT_QUERY,
      variables: {},
    });
    expect(res).toMatchSnapshot();
  });

  it('should not find a user by email if it does not exist', async () => {
    const res = await server.executeOperation({
      query: FIND_CONTACT_QUERY,
      variables: { email: 'fghjkbnjyghm' },
    });
    expect(res).toMatchSnapshot();
  });
});

export {};
