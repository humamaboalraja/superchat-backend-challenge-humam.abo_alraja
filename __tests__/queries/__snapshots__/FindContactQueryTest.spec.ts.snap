// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test find contact query should find a contact given their email 1`] = `
Object {
  "data": Object {
    "findContact": Object {
      "email": "humam@company.com",
      "name": "Humam Abo Alraja",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test find contact query should not find a contact if no arguments are passed in 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Contact.name.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test find contact query should not find a user by email if it does not exist 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Contact.name.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
